File: app/src/main/java/com/jwu5/giphyapp/viewholder/GiphyViewHolder.java
package com.jwu5.giphyapp.viewholder;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.jwu5.giphyapp.adapters.GiphyRecyclerViewAdapter;
import com.jwu5.giphyapp.R;
import com.jwu5.giphyapp.network.model.GiphyModel;

/**
 * Created by Jiawei on 8/16/2017.
 */

public class GiphyViewHolder extends RecyclerView.ViewHolder implements ViewHolderView{
    private ImageView mItemImageView;
    private ImageButton mFavoriteButton;
    private ProgressBar mProgressBar;
    private Context mContext;
    private GiphyRecyclerViewAdapter mAdapter;
    private String TAG;

    protected ViewHolderPresenter mViewHolderPresenter;

    public GiphyViewHolder(View itemView, Context context, GiphyRecyclerViewAdapter adapter, String tag) {
        super(itemView);
        mContext = context;
        mItemImageView = (ImageView)itemView.findViewById(R.id.Giphy_image_view);
        mProgressBar = (ProgressBar) itemView.findViewById(R.id.progress_bar);
        mFavoriteButton = (ImageButton)itemView.findViewById(R.id.Giphy_image_button);
        mAdapter = adapter;

        TAG = tag;
        mViewHolderPresenter = new ViewHolderPresenter(this, mContext);
    }

    public void bindGiphyItem(final GiphyModel item) {
        Glide
                .with(mContext)
                .load(item.getUrl())
                .listener(new RequestListener<String, GlideDrawable>() {
                    @Override
                    public boolean onException(Exception e, String model, Target<GlideDrawable> target, boolean isFirstResource) {
                        mProgressBar.setVisibility(View.GONE);
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
                        mProgressBar.setVisibility(View.GONE);
                        return false;
                    }
                })
                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                .into(mItemImageView);

        mViewHolderPresenter.setFavoriteIcon(item);

        mFavoriteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mViewHolderPresenter.saveToFavorites(item);
            }
        });
    }

    @Override
    public void setButtonColor(int color) {
        mFavoriteButton.setColorFilter(color);
    }

    @Override
    public void removeItemFromView(GiphyModel item) {
        mAdapter.removeItem(item);
    }

    @Override
    public void updateAdapter() {
        mAdapter.notifyItemChanged(getAdapterPosition());
    }

    @Override
    public String getTag() {
        return TAG;
    }

}

File: app/src/main/java/com/jwu5/giphyapp/viewholder/GiphyViewHolder.kt
package com.jwu5.giphyapp.viewholder

import android.content.Context
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.ProgressBar

import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.GlideDrawable
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.jwu5.giphyapp.adapters.GiphyRecyclerViewAdapter
import com.jwu5.giphyapp.R
import com.jwu5.giphyapp.network.model.GiphyModel

/**
 * Created by Jiawei on 8/16/2017.
 */

class GiphyViewHolder(itemView: View, private val mContext: Context, private val mAdapter: GiphyRecyclerViewAdapter, override val tag: String) : RecyclerView.ViewHolder(itemView), ViewHolderView {
    private val mItemImageView: ImageView
    private val mFavoriteButton: ImageButton
    private val mProgressBar: ProgressBar
    private val WHITE = Color.WHITE
    private val PINK = Color.argb(255, 255, 102, 153)

    protected var mViewHolderPresenter: ViewHolderPresenter

    init {
        mItemImageView = itemView.findViewById(R.id.Giphy_image_view) as ImageView
        mProgressBar = itemView.findViewById(R.id.progress_bar) as ProgressBar
        mFavoriteButton = itemView.findViewById(R.id.Giphy_image_button) as ImageButton
        mViewHolderPresenter = ViewHolderPresenter(this, mContext)
    }

    fun bindGiphyItem(item: GiphyModel) {
        Glide.with(mContext).load(item.url).listener(object : RequestListener<String, GlideDrawable> {
            override fun onException(e: Exception, model: String, target: Target<GlideDrawable>, isFirstResource: Boolean): Boolean {
                mProgressBar.visibility = View.GONE
                return false
            }

            override fun onResourceReady(resource: GlideDrawable, model: String, target: Target<GlideDrawable>, isFromMemoryCache: Boolean, isFirstResource: Boolean): Boolean {
                mProgressBar.visibility = View.GONE
                return false
            }
        }).diskCacheStrategy(DiskCacheStrategy.SOURCE).into(mItemImageView)

        mViewHolderPresenter.setFavoriteIcon(item)

        mFavoriteButton.setOnClickListener { mViewHolderPresenter.saveToFavorites(item) }
    }

    override fun getInFavoritesColor(): Int {
        return PINK;
    }

    override fun getNotInFavoritesColor(): Int {
        return WHITE;
    }

    override fun setButtonColor(color: Int) {
        mFavoriteButton.setColorFilter(color)
    }

    override fun removeItemFromView(item: GiphyModel) {
        mAdapter.removeItem(item)
    }

    override fun updateAdapter() {
        mAdapter.notifyItemChanged(adapterPosition)
    }

}
